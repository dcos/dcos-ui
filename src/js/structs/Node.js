import Item from "./Item";
import TaskStates from "../../../plugins/services/src/js/constants/TaskStates";
import UnitHealthUtil from "../utils/UnitHealthUtil";
import { findNestedPropertyInObject } from "../utils/Util";

class Node extends Item {
  getID() {
    return this.get("id");
  }

  getServiceIDs() {
    return this.get("framework_ids");
  }

  isActive() {
    return this.get("active");
  }

  getDomain() {
    return this.get("domain");
  }

  getRegionName() {
    const nodeRegionName = findNestedPropertyInObject(
      this.getDomain(),
      "fault_domain.region.name"
    );

    return nodeRegionName || "N/A";
  }

  getZoneName() {
    const nodeZoneName = findNestedPropertyInObject(
      this.getDomain(),
      "fault_domain.zone.name"
    );

    return nodeZoneName || "N/A";
  }

  getUsageStats(resource) {
    const total = (this.get("resources") || {})[resource];
    const value = (this.get("used_resources") || {})[resource];

    if (total === undefined || value === undefined) {
      return { percentage: 0, total: 0, value: 0 };
    }

    const percentage = Math.round((100 * value) / Math.max(1, total));

    return { percentage, total, value };
  }

  getHostName() {
    return this.get("hostname");
  }

  // Below is Component Health specific API
  // http://schema.dcos/system/health/node
  getHealth() {
    return UnitHealthUtil.getHealth(this.get("health"));
  }

  getOutput() {
    if (typeof this.get("output") === undefined) {
      return "N/A";
    }

    return this.get("output") || "OK";
  }

  sumTaskTypesByState(state) {
    let sum = 0;

    Object.keys(TaskStates).forEach(function(taskType) {
      if (
        TaskStates[taskType] &&
        TaskStates[taskType].stateTypes.indexOf(state) !== -1
      ) {
        // Make sure there's a value
        if (this[taskType]) {
          sum += this[taskType];
        }
      }
    }, this);

    return sum;
  }

  getResources() {
    return (
      this.get("used_resources") || {
        cpus: 0,
        mem: 0,
        gpus: 0,
        disk: 0
      }
    );
  }

  isPublic() {
    return (
      findNestedPropertyInObject(this.get("attributes"), "public_ip") === "true"
    );
  }

  getIp() {
    return this.get("host_ip") || this.getHostName();
  }

  getPublicIps() {
    if (!this.get("network")) {
      return [];
    }

    return this.get("network").public_ips || [];
  }
}

module.exports = Node;
