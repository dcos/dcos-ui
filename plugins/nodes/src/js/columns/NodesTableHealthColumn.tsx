import * as React from "react";
import UnitHealthUtil from "#SRC/js/utils/UnitHealthUtil";
import Node from "#SRC/js/structs/Node";
// TODO: DCOS-39079
// import { IWidthArgs as WidthArgs } from "@dcos/ui-kit/packages/table/components/Column";
import { IWidthArgs as WidthArgs } from "#PLUGINS/nodes/src/js/types/IWidthArgs";
import { SortDirection } from "plugins/nodes/src/js/types/SortDirection";
import { TextCell } from "@dcos/ui-kit";

export function healthRenderer(data: Node): React.ReactNode {
  const health = data.getHealth();
  return (
    <TextCell>
      <span className={health.classNames}>{health.title}</span>
    </TextCell>
  );
}

export function healthSorter(
  data: Node[],
  sortDirection: SortDirection
): Node[] {
  const sortedData = data.sort(UnitHealthUtil.getHealthSortFunction);
  return sortDirection === "ASC" ? sortedData : sortedData.reverse();
}

export function healthSizer(_args: WidthArgs): number {
  // TODO: DCOS-38827
  return 120;
}
