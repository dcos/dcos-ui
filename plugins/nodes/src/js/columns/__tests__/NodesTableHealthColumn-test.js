import TestRenderer from "react-test-renderer";
import Node from "#SRC/js/structs/Node";

import NA from "./fixtures/node-without-health.json";
import Healthy from "./fixtures/node-healthy.json";
import Unhealthy from "./fixtures/node-unhealthy.json";
import Warn from "./fixtures/node-warn.json";

import { healthRenderer } from "../NodesTableHealthColumn";

describe("NodesTableHealthColumn", () => {
  describe("renderer", () => {
    const testCases = { NA, Healthy, Unhealthy, Warn };
    for (const [statusName, data] of Object.entries(testCases)) {
      it(`renders with health status '${statusName}'`, () => {
        expect(
          TestRenderer.create(healthRenderer(new Node(data))).toJSON()
        ).toMatchSnapshot();
      });
    }
  });
});
