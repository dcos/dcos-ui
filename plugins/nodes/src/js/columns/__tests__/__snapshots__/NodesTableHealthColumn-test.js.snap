// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodesTableHealthColumn renderer renders with health status 'Healthy' 1`] = `
<div
  className="css-1c5c0a1"
>
  <span
    className="text-success"
  >
    Healthy
  </span>
</div>
`;

exports[`NodesTableHealthColumn renderer renders with health status 'NA' 1`] = `
<div
  className="css-1c5c0a1"
>
  <span
    className="text-mute"
  >
    N/A
  </span>
</div>
`;

exports[`NodesTableHealthColumn renderer renders with health status 'Unhealthy' 1`] = `
<div
  className="css-1c5c0a1"
>
  <span
    className="text-danger"
  >
    Unhealthy
  </span>
</div>
`;

exports[`NodesTableHealthColumn renderer renders with health status 'Warn' 1`] = `
<div
  className="css-1c5c0a1"
>
  <span
    className="text-warning"
  >
    Warning
  </span>
</div>
`;
